module3/lessons/testing_tools_for_api_consumption.md

```diff
@@ -0,0 +1,468 @@
+
+## Setup
+
+First, add `gem "vcr"` to the `:test` block of your Gemfile and `bundle install`.
+
+Then, add this at the bottom of your `rails_helper`:
+
+*spec/rails_helper.rb*
+
+```ruby
+VCR.configure do |config|
+  config.cassette_library_dir = "spec/fixtures/vcr_cassettes"
+  config.hook_into :webmock
+end
+```
+
+In the first line of the block, we tell VCR where we want to store the the `cassettes`. We are making use of the `spec/fixtures` folder we already created.
+
+The second line tells VCR what library it should use for intercepting these requests, which will be WebMock. So we are still using WebMock, but VCR is adding additional functionality for recording responses.
+
+Let's reset our test in *spec/requests/api/v1/images/image_request_spec.rb* by removing the lines we added to use webmock.
+
+Now, let’s run our test.
+
+```bash
+$ bundle exec rspec spec/requests/api/v1/images/image_request_spec.rb
+```
+
+You’re going to see a big error, but the important part is:
+
+```ruby
+# --- Caused by: ---
+     # VCR::Errors::UnhandledHTTPRequestError:
+     #
+     #
+     #   ================================================================================
+     #   An HTTP request has been made that VCR does not know how to handle:
+     #      GET https://api.pexels.com/v1/search?query=The%20Beatles
+```
+
+This means that it’s working. 
+
+## Stubbing the Request
+
+In order to use VCR, we wrap our test in a `VCR.use_cassette` block:
+
+*spec/requests/api/v1/images/image_request_spec.rb*
+
+```ruby
+require "rails_helper"
+
+RSpec.describe "Images Endpoint" do
+  describe "happy path" do
+    it "can retrieve an image for a specific artist specific artist" do
+      VCR.use_cassette("beatles_artist_query") do
+
+        get "/api/v1/images?artist=The%20Beatles"
+
+        expect(response).to be_successful
+        json = JSON.parse(response.body, symbolize_names: true)
+
+        expect(json[:data][:id]).to be_nil
+        expect(json[:data][:type]).to eq("image")
+        expect(json[:data][:attributes]).to have_key(:image_url)
+        expect(json[:data][:attributes]).to have_key(:photographer)
+        expect(json[:data][:attributes]).to have_key(:photographer_url)
+        expect(json[:data][:attributes]).to have_key(:alt_text)
+      end
+    end
+  end
+end
+```
+
+The string we passed to `use_cassette` is an identifier for the cassette, so it doesn't really matter what you pass it.
+
+Run your tests and they should be passing. If you look under `spec/fixtures/vcr_cassettes` you should see a `.yml` file that contains your recorded response.
+
```

Comment:
We might want to point out that at this point, they are no longer using the .json file that contains a copy of the response.  I could see students misunderstanding that the cassette replaces the idea of the manually stubbed response.